contracts/axelar-gateway/tests/test.rs

```diff
@@ -490,3 +492,346 @@ fn upgrade_unauthorized() {
+    let proof = generate_proof(&env, msg_hash, signers);
+
+    // should panic, proof is for different message hash
+    // NOTE: panic occurs in std function, cannot handle explicitly
+    client.approve_messages(&messages, &proof);
+}
+
+#[test]
+fn fail_validate_proof_empty_signatures() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let msg_hash: BytesN<32> = BytesN::random(&env);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers.clone());
+
+    // Modify signatures to make them invalid
+    let mut new_signers = Vec::new(&env);
+    for signer in proof.signers.iter() {
+        new_signers.push_back(ProofSigner {
+            signer: signer.signer,
+            signature: ProofSignature::Unsigned,
+        });
+    }
+    proof.signers = new_signers;
+
+    let new_signers = generate_signers_set(&env, 5, signers.domain_separator.clone());
+
+    assert_contract_err!(
+        client.try_rotate_signers(&new_signers.signers, &proof, &true),
+        ContractError::InvalidSignatures
+    );
+}
+
+#[test]
+fn fail_validate_proof_threshold_not_met() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let mut total_weight = 0u128;
+
+    let (message, _) = generate_test_message(&env);
+    let messages = vec![&env, message.clone()];
+    let msg_hash = get_approve_hash(&env, messages.clone());
+
+    // let msg_hash: BytesN<32> = BytesN::random(&env);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    // Modify signatures to make them invalid
+    let mut new_signers = Vec::new(&env);
+    for ProofSigner { signer, signature } in proof.signers {
+        total_weight += signer.weight;
+
+        if total_weight < proof.threshold {
+            new_signers.push_back(ProofSigner { signer, signature });
+        } else {
+            new_signers.push_back(ProofSigner {
+                signer,
+                signature: ProofSignature::Unsigned,
+            });
+        }
+    }
+    proof.signers = new_signers;
+
+    // should panic, all signatures are valid but total weight is below threshold
+    assert_contract_err!(
+        client.try_approve_messages(&messages, &proof),
+        ContractError::InvalidSignatures
+    );
+}
+
+#[test]
+fn fail_validate_proof_invalid_signer_set() {
+    let (env, signers, client) = setup_env(randint(0, 10), randint(1, 10));
+
+    let new_signers = generate_signers_set(&env, randint(1, 10), signers.domain_separator.clone());
+
+    let msg_hash: BytesN<32> = BytesN::random(&env);
+    let mut proof = generate_proof(&env, msg_hash.clone(), signers);
+
+    let new_proof = generate_proof(&env, msg_hash.clone(), new_signers.clone());
+
+    proof.signers = new_proof.signers;
+
+    // validate_proof should panic, signatures do not match signers
+    assert_contract_err!(
+        client.try_rotate_signers(&new_signers.signers, &proof, &true),
+        ContractError::InvalidSignersHash
+    );
```

Comment:
```suggestion
    let invalid_proof = generate_proof(&env, msg_hash.clone(), new_signers.clone());

    assert_contract_err!(
        client.try_rotate_signers(&new_signers.signers, &invalid_proof, &true),
        ContractError::InvalidSignersHash
    );
```
avoid mutation, and use descriptive names so comments aren't needed